nengo/builder/optimizer.py

```diff
@@ -11,386 +10,707 @@
+        from scipy.sparse import bsr_matrix
+        assert bsr_matrix
+    except ImportError:
+        warnings.warn(
+            "Skipping some optimization steps because SciPy is not installed. "
+            "Installing SciPy may result in faster simulations.")
+    except ValueError:
+        return False
+    return (_SigMerger.check([op1.X, op2.X]) and
+            _SigMerger.check([op1.Y, op2.Y]) and
+            op1.A.shape == op2.A.shape)
+
+
+@_OpMerger.register_sig_checker(operator.DotInc)
+def dotinc_sig_checker(op, tomerge):
+    none_shared = (default_sig_checker(op, tomerge) and
+                   len(set(o.X for o in tomerge.ops)) == len(tomerge.ops))
+    all_x_shared = (
+        not any(set((op.X, op.A, op.Y)).intersection((o.A, o.Y))
+                for o in tomerge.ops) and
+        all(op.X is o.X for o in tomerge.ops))
+    return none_shared or all_x_shared
+
+
+@_OpMerger.register_merger(operator.DotInc)
+def dotinc_merge(ops):
+    # Simple merge if all X are the same.
+    if all(o.X is ops[0].X for o in ops):
+        A, A_sigr = _SigMerger.merge([o.A for o in ops])
+        Y, Y_sigr = _SigMerger.merge([o.Y for o in ops])
+        return operator.DotInc(A, ops[0].X, Y), merge_dicts(A_sigr, Y_sigr)
+    assert all(o1.X is not o2.X
+               for i, o1 in enumerate(ops) for o2 in ops[i+1:])
+
+    # BSR merge if X differ
+    X, X_sigr = _SigMerger.merge([o.X for o in ops])
+    Y, Y_sigr = _SigMerger.merge([o.Y for o in ops])
+
+    # Construct sparse A representation
+    data = np.array([o.A.initial_value for o in ops])
+    if data.ndim == 1:
+        data = data.reshape((data.size, 1, 1))
+    elif data.ndim == 2:
+        data = data.reshape(data.shape + (1,))
+    indptr = np.arange(len(ops) + 1, dtype=int)
+    indices = np.arange(len(ops), dtype=int)
+    name = 'bsr_merged<{first}, ..., {last}>'.format(
+        first=ops[0].A.name, last=ops[-1].A.name)
+    readonly = all([o.A.readonly for o in ops])
+    A = Signal(data, name=name, readonly=readonly)
+    A_sigr = {}
+    for i, s in enumerate([o.A for o in ops]):
+        A_sigr[s] = Signal(data[i], name="%s[%i]" % (s.name, i), base=A)
+        assert np.all(s.initial_value == A_sigr[s].initial_value)
+        assert s.shape == A_sigr[s].shape or (
+            s.shape == () and A_sigr[s].shape == (1, 1))
+
+    reshape = operator.reshape_dot(
+        ops[0].A.initial_value, ops[0].X.initial_value,
+        ops[0].Y.initial_value, tag=ops[0].tag)
+    return (
+        operator.BsrDotInc(
+            A, X, Y, indices=indices, indptr=indptr, reshape=reshape),
+        merge_dicts(X_sigr, Y_sigr, A_sigr))
+
+
+@_OpMerger.register_op_checker(SimNeurons)
+def simneurons_checker(op1, op2):
+    return (op1.neurons == op2.neurons and
+            all(_SigMerger.check(s) for s in
+                zip(op1.all_signals, op2.all_signals)))
+
+
+def _gather(ops, key):
+    return [getattr(o, key) for o in ops]
```

Comment:
I was slightly surpresit to find a non-checker/non-merger function here, it seems a bit out of place here even though it is only used in `simneurons_merger`. Maybe move `_gather` into `simneurons_merger`?